storage/src/main/java/io/delta/storage/LogStore.java

```diff
@@ -72,47 +77,53 @@ public LogStore(Configuration initHadoopConf) {
    * Load the given file and return an `Iterator` of lines, with line breaks removed from each line.
    * Callers of this function are responsible to close the iterator if they are done with it.
    *
+   * @throws IOException if there's an issue resolving the FileSystem
    * @since 1.0.0
    */
-  public abstract CloseableIterator<String> read(Path path, Configuration hadoopConf);
+  public abstract CloseableIterator<String> read(Path path, Configuration hadoopConf) throws IOException;
 
   /**
    * :: DeveloperApi ::
    *
    * Write the given `actions` to the given `path` with or without overwrite as indicated.
-   * Implementation must throw [[java.nio.file.FileAlreadyExistsException]] exception if the file
+   * Implementation must throw {@link java.nio.file.FileAlreadyExistsException} exception if the file
    * already exists and overwrite = false. Furthermore, if isPartialWriteVisible returns false,
    * implementation must ensure that the entire file is made visible atomically, that is,
    * it should not generate partial files.
    *
+   * @throws IOException if there's an issue resolving the FileSystem
+   * @throws FileAlreadyExistsException if the file already exists and overwrite is false
    * @since 1.0.0
    */
   public abstract void write(
       Path path,
       Iterator<String> actions,
       Boolean overwrite,
-      Configuration hadoopConf) throws FileAlreadyExistsException;
+      Configuration hadoopConf) throws IOException;
 
   /**
    * :: DeveloperApi ::
    *
    * List the paths in the same directory that are lexicographically greater or equal to
    * (UTF-8 sorting) the given `path`. The result should also be sorted by the file name.
    *
+   * @throws IOException if there's an issue resolving the FileSystem
+   * @throws FileAlreadyExistsException if {@code path} directory can't be found
    * @since 1.0.0
    */
   public abstract Iterator<FileStatus> listFrom(
       Path path,
-      Configuration hadoopConf) throws FileNotFoundException;
+      Configuration hadoopConf) throws IOException;
```

Comment:
Technically, this is a minor breaking change for Java LogStore. For example,
```
public class FooLogStore extends HadoopFileSystemLogStore {
  @Override
  public Iterator<FileStatus> listFrom(Path path, Configuration hadoopConf) throws FileNotFoundException {
    Iterator<FileStatus> iter1 = super.listFrom(path, hadoopConf);
    Iterator<FileStatus> iter2 = // list from somewhere else
    return merge(iter1, iter2);
  }
}
```
This will be broken in the next release as `IOException` is not caught.

But I'm okay with this minor breaking changes given this is the developer API and the LogStore API is still in the early stage. @tdas thoughts?

Reply:
Yup, I do prefer this breaking change.

Reply:
can someone explain succinctly why this exception type change is needed?

Reply:
TLDR: java implementation calls `path.getFileSystem(hadoopConf)`. This throws an IOException, not a FileNotFoundException. So, it makes sense to update the interface.

Detail:

In the initial version of `LogStore.java` interface, we pretty much just copied the `LogStore.scala` interface as best we could. The scala implementation for `listFrom` includes:
```
if (!fs.exists(path.getParent)) {
      throw new FileNotFoundException(s"No such file or directory: ${path.getParent}")
    }
```

so, we added a `throws FileNotFoundException` to the interface. Makes sense.

But, when we actually go to **implement** the interface, every single `path.getFileSystem(hadoopConf)` risks throwing an IOException. We didn't think about this when we made the interface. We didn't worry about this in Scala interface.